src/lib/CollisionPrevention/CollisionPrevention.cpp

```diff
@@ -69,104 +65,83 @@ bool CollisionPrevention::initializeSubscriptions(SubscriptionArray &subscriptio
+void CollisionPrevention::calculate_constrained_setpoint(const float max_acc, Vector2f &setpoint)
 {
+	const obstacle_distance_s &obstacle_distance = _sub_obstacle_distance->get();
 
-	collision_constraints_s	constraints;	/**< collision constraints message */
-
-	//fill in values
-	constraints.timestamp = hrt_absolute_time();
-	constraints.constraints_normalized_x[0] = _move_constraints_x_normalized(0);
-	constraints.constraints_normalized_x[1] = _move_constraints_x_normalized(1);
-	constraints.constraints_normalized_y[0] = _move_constraints_y_normalized(0);
-	constraints.constraints_normalized_y[1] = _move_constraints_y_normalized(1);
-
-	constraints.original_setpoint[0] = original_setpoint(0);
-	constraints.original_setpoint[1] = original_setpoint(1);
-	constraints.adapted_setpoint[0] = adapted_setpoint(0);
-	constraints.adapted_setpoint[1] = adapted_setpoint(1);
+	//The maximum velocity formula contains a square root, therefore the whole calculation is done with squared norms.
+	//that way the root does not have to be calculated for every range bin but once at the end.
+	Vector2f setpoint_sqrd = setpoint * setpoint.norm();
 
-	// publish constraints
-	if (_constraints_pub != nullptr) {
-		orb_publish(ORB_ID(collision_constraints), _constraints_pub, &constraints);
+	//Limit the deviation of the adapted setpoint from the originally given joystick input (slightly less than 90 degrees)
+	float max_slide_angle_rad = 1.2f;
 
-	} else {
-		_constraints_pub = orb_advertise(ORB_ID(collision_constraints), &constraints);
-	}
-}
-
-void CollisionPrevention::update_range_constraints()
-{
-	const obstacle_distance_s &obstacle_distance = _sub_obstacle_distance->get();
-
-	if (hrt_elapsed_time(&obstacle_distance.timestamp) < RANGE_STREAM_TIMEOUT_US) {
-		float max_detection_distance = obstacle_distance.max_distance / 100.0f; //convert to meters
+	if (hrt_elapsed_time(&obstacle_distance.timestamp) < RANGE_STREAM_TIMEOUT_US && setpoint.norm() > 0.001f) {
 
 		int distances_array_size = sizeof(obstacle_distance.distances) / sizeof(obstacle_distance.distances[0]);
 
 		for (int i = 0; i < distances_array_size; i++) {
+
 			//determine if distance bin is valid and contains a valid distance measurement
 			if (obstacle_distance.distances[i] < obstacle_distance.max_distance &&
 			    obstacle_distance.distances[i] > obstacle_distance.min_distance && i * obstacle_distance.increment < 360) {
 				float distance = obstacle_distance.distances[i] / 100.0f; //convert to meters
 				float angle = math::radians((float)i * obstacle_distance.increment);
 
-				//calculate normalized velocity reductions
-				float vel_lim_x = (max_detection_distance - distance) / (max_detection_distance - _param_mpc_col_prev_d.get()) * cos(
-							  angle);
-				float vel_lim_y = (max_detection_distance - distance) / (max_detection_distance - _param_mpc_col_prev_d.get()) * sin(
-							  angle);
+				//max velocity squared in current bin direction
```

Comment:
Any particular reason for choosing half the limit?

Reply:
We want to make sure we are well inside of the vehicle dynamics for something critical like avoidance. Also note we are assuming infinite jerk, so this gives us enough margin to make that assumption not too bad.

Reply:
Could this use the maximum jerk parameter instead? FYI @bresch

Reply:
@jlecoeur it would need to use bot max accel _and_ max jerk, since accel gets saturated pretty quickly. The math gets a bit more complicated due to the discontinuous jerk, so we decided to stick with the simple version.

If there is a fairly lightweight formula we could use, that would be great. Keep in mind this runs up to 60 times (depending on how much laser data is populated).

Reply:
@jkflying Sure, it would need to limit the velocity based on max acceleration and max jerk. The approximation you make looks reasonable to me. Is the maths described somewhere, is this also in the Master thesis?

Reply:
@jlecoeur I just added a term which takes into account the acceleration ramp up time/distance considering the maximum jerk limit. There is no official documentation for this yet, I tried to visualize the principle a bit in the graphic for those who are interested. Once this gets in, I'll make sure to provide the documentation somewhere in the dev guide, I guess.

Reply:
Awesome! I have a small remark regarding the ramp-up which I made in the code.